experiments/ppo_gridnet.py

```diff
@@ -479,39 +482,53 @@ def get_value(self, x):
-                        preset_trueskill_step_df_clone["step"] = model_global_step
-                        trueskill_step_df = trueskill_step_df.append(preset_trueskill_step_df_clone, ignore_index=True)
-                        wandb.log({"trueskill_step": wandb.Table(dataframe=trueskill_step_df)})
+                        "--output-path",
+                        f"runs/{experiment_name}/{global_step}.csv",
+                    ],
+                    f"models/{experiment_name}/{global_step}.pt",
+                    f"runs/{experiment_name}/{global_step}.csv",
+                ]
+            ]
+            print(f"Queued models/{experiment_name}/{global_step}.pt")
+
+        while len(eval_queue) > 0:
+            # always start the max-eval-workers processes
+            for idx in range(len(eval_queue[: args.max_eval_workers])):
+                if type(eval_queue[idx][0]) == list:
+                    eval_queue[idx][0] = subprocess.Popen(eval_queue[idx][0])
+                    print(f"Evaluating {eval_queue[idx][1]}")
+
+            if eval_queue[0][0].poll() is not None:
+                # read and clean up
+                league = pd.read_csv(eval_queue[0][2], index_col="name")
+                model_path = eval_queue[0][1]
+                eval_queue = eval_queue[1:]
+                assert model_path in league.index
+                model_global_step = int(model_path.split("/")[-1][:-3])
+                writer.add_scalar("charts/trueskill", league.loc[model_path]["trueskill"], model_global_step)
+                print(f"global_step={model_global_step}, trueskill={league.loc[model_path]['trueskill']}")
+
+                # table visualization logic
+                if args.prod_mode:
+                    trueskill_data = {
+                        "name": league.loc[model_path].name,
+                        "mu": league.loc[model_path]["mu"],
+                        "sigma": league.loc[model_path]["sigma"],
+                        "trueskill": league.loc[model_path]["trueskill"],
+                    }
+                    trueskill_df = trueskill_df.append(trueskill_data, ignore_index=True)
+                    wandb.log({"trueskill": wandb.Table(dataframe=trueskill_df)})
+                    trueskill_data["type"] = "training"
+                    trueskill_data["step"] = model_global_step
+                    trueskill_step_df = trueskill_step_df.append(trueskill_data, ignore_index=True)
+                    preset_trueskill_step_df_clone = preset_trueskill_step_df.copy()
+                    preset_trueskill_step_df_clone["step"] = model_global_step
+                    trueskill_step_df = trueskill_step_df.append(preset_trueskill_step_df_clone, ignore_index=True)
+                    wandb.log({"trueskill_step": wandb.Table(dataframe=trueskill_step_df)})
+
+            # if the training is not done, don't wait all eval processes to finish
+            if not update == args.num_updates:
```

Comment:
If I'm reading this correctly, the idea is to run subprocess while training still happening + wait for all of them to finish if when training is done. Right? Can we do this with threading pool? (`concurrent.futures.ThreadPoolExecutor`):

1. Set up thread pool with max workers `args.max_eval_workers`
2. Submit to the pool new evaluation when ready (lambda to open subprocess and wait for it to finish)
3. Use `add_done_callback` to execute CSV update
4. Join at the end to make sure all submitted tasks are finished

Should be conceptually simpler. WDYT? I'll be glad to help with the implementation

Reply:
I think it will indeed be cleaner code, although I had some difficulties running it. I tried to run https://gist.github.com/vwxyzjn/36375efb035f1508ba0c44e7fd15afb9

and had the following results:

If you could help look into this that will be great :)

```
Model's total parameters: 838191
SPS: 800
exception calling callback for <Future at 0x7f52ac090b80 state=finished raised TypeError>
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/concurrent/futures/process.py", line 208, in _sendback_result
    result_queue.put(_ResultItem(work_id, result=result,
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/multiprocessing/queues.py", line 372, in put
    obj = _ForkingPickler.dumps(obj)
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle '_thread.lock' object
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/concurrent/futures/_base.py", line 329, in _invoke_callbacks
    callback(self)
  File "/home/costa/Documents/go/src/github.com/gym-microrts/experiments/ppo_gridnet.py", line 493, in write_trueskill
    print(future.result())
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/concurrent/futures/_base.py", line 438, in result
    return self.__get_result()
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
TypeError: cannot pickle '_thread.lock' object
match_qualities[:3] [[<AI: 🤖 guidedRojoA3N with N(26.228, 0.979)>, 0.5706042609450162], [<AI: 🤖 rojo with N(26.345, 1.056)>, 0.569372689002722], [<AI: 🤖 naiveMCTSAI with N(26.798, 1.042)>, 0.5656072235262297]]
the match up is (🤖 models/MicroRTSGridModeVecEnv__ppo_gridnet__1__1643476830/6144.pt with N(25.0, 8.333), 🤖 rojo with N(26.345, 1.056)), quality is 0.5694
rojo wins models/MicroRTSGridModeVecEnv__ppo_gridnet__1__1643476830/6144.pt
SPS: 774
exception calling callback for <Future at 0x7f52ac0248b0 state=finished raised TypeError>
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/concurrent/futures/process.py", line 208, in _sendback_result
    result_queue.put(_ResultItem(work_id, result=result,
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/multiprocessing/queues.py", line 372, in put
    obj = _ForkingPickler.dumps(obj)
  File "/home/costa/.pyenv/versions/3.9.5/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle '_thread.lock' object
```

Reply:
That's interesting 🤔 I will absolutely take a look today!

Reply:
Regarding

> is still affected by GIL

It definitely is. But `subprocess.Popen` creates another process, and the current process only tracks I/O operations on open descriptors. Which takes trivial amount of compute (and, we did this anyways within the same GIL context when calling `.poll`).

Reply:
Oh I see, you are using thread pool to do `subprocess.Popen`. This makes sense to me.

Reply:
Github doesn't let me to attach patch files in comments 😭 First pass is ready, seems to work okay. I'm playing around trying to find an easy way to run DataFrame update only in main thread (initial implementation uses a lock for this and it's somewhat annoying)

Reply:
Oh really I didn’t find any lock in #55, I assume you have since removed it? 👀

Reply:
Yeah, I only PR'ed the second version of the implementation